Add SLO Lifecycle ADR

lisa opened the pull request:
Add the SLO Lifecycle document.

There are some places here that will need to be edited after a merge since there are references to RACI document that will want to link to that document, but that's not ready yet (and also links to this document).

For SIGSRE-59.

Signed-off-by: Lisa Seelye <lisa@users.noreply.github.com>

lisa committed:
Add SLO Lifecycle ADR

Signed-off-by: Lisa Seelye <lisa@users.noreply.github.com>

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+
+## Proposed Architecture
+
+Add Service Level Objectives (SLOs) as a feature to services with its own distinct lifecycle described herein. The PM will be accountable for the SLO feature.
```
We got some feedback that using the word feature would not work out the way I thought it would, recommend not using this word.

lisa replied:
Good catch. I will make the change.

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+
+With an accepted proposal from the Research Phase, an owner can be found to implement the changes. Not all of these changes are necessarily made by software engineers. Certainly changes to instrument the service’s codebase is in the domain of the software engineer, but changing team policies and practices is more likely to be the domain of the engineering lead. Remember that the work isn’t intend ...
+
+The owner will likely be an engineer on the service's team who may work in concert with an appropriate SRE IC to affect changes to the service and related ecosystem.
```
s/IC//

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+
+Part of healthy SLO adoption is the periodic review of the service's SLO needs. These review tasks may include:
+
+* Do the service's SLOs still measure the right thing(s) to measure customer happiness and service reliability?
```
Add a bullet here that mentions the customer experience.  Have we gotten any direct or telemetry-inferred feedback that indicates we should change something?

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+These steps can be used to evaluate the SLOs themselves, but there's also review on the targets set by the service's SLO(s):
+
+* Are the service's SLOs being met?
+  * If they're not, what are common reasons why not?
```
And who's job is it to fix?  e.g. do we have upstream quality issues that could be caught by new tests?

lisa replied:
I will add a new section for this.

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+
+#### Sprint Planning Hooks
+
+A primary goal of Service Level Objectives is to inform the natural cycle of work for product feature teams by adding an additional signal for use in planning work. The new signal is the service's remaining Error Budget.
```
I feel like this sentence is so important that we should either move it to the top of this doc, or copy it to the top of the doc and keep it here, too.

lisa replied:
Agreed. I copied it to the Problem Statement section to serve as a kind of preamble.

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+## Challenges
+
+* Teams may desire to adopt the guidelines in this document but lack the data to do so because they don't have any SLO, or a way to measure the remaining error budget.
+* Teams may face encounter difficulties scheduling reliability work by those who deprioritize it in favour of "feature work."
```
Here I think we should take some of what @mh21 mentioned his team does, where they have an "informing" phase for an SLO which allows them time to calibrate, before graduating it to "enforcing".  Have I got that right, @mh21 ?

jeremyeder replied:
Err, sorry or was this @twaugh group?

lisa replied:
I can't recall who, but that's a great idea! I'll add a section about a "permissive phase."

jeremyeder commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,131 @@
+
+* The existence of an SLO
+* Reporting for SLO compliance and error budget exhaustion
+* Willingness to prioritize reliability work over feature work when necessary
```
here again, use the word customer or customer experience, it's not reliability for reliabilities sake. its because we're proxying customer experience through SLOs and these generally focus on reliability (but not exclusively).

lisa committed:
Address PR feedback

Signed-off-by: Lisa Seelye <lisa@users.noreply.github.com>

dastergon commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,153 @@
+
+### Permissive SLO Phase
+
+Teams may find that adopting this SLO lifecycle challenging for any number of reasons, whether they're listed in the [challenges](#challenges) section or not. For teams that want adopt a lightweight approach to the SLO lifecycle, a "permissive phase" could be helpful.
```
We could also mention here that we can also have aspirational but realistic SLOs and actual SLOs together.

lisa replied:
Don't aspirational SLOs aim to become "actual" SLOs, though?

dastergon replied:
Yeap, just as a concept though, they can be implemented to help calibrate your actual SLOs better

lisa replied:
@dastergon I've had a think and maybe we leave that out of the lifecycle doc here and give the idea a home in the kind of slo cookbook, since I know we have teams that chose that path.

jeremyeder replied:
Let's table the aspirational SLOs for the moment only because we can add it later to either this, or the cookbook once we have a worked example.

lisa commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,153 @@
+
+* Teams may desire to adopt the guidelines in this document but lack the data to do so because they don't have any SLO, or a way to measure the remaining error budget. 
+* Teams may face encounter difficulties scheduling reliability work by those who deprioritize it in favour of "feature work."
+  * Related, teams may not have the throughput to address reliability work above "keeping the lights on" work. (See also toil management.)
```
"feature work" or reliability work. Which quotes go where? Be consistent.

lisa commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,153 @@
+# ADR-00000 Template
```
@lisa this should be ADR-00005 SLO Lifecycle

lisa commented on ADRs/RH/SIG-SRE/ADR-00005 SLO Lifecycle.md:
```diff
@@ -0,0 +1,153 @@
+## Dependencies
+
+* The existence of an SLO
+* Reporting for SLO compliance and error budget exhaustion
```
End with a `.`

lisa committed:
PR Feedback

Signed-off-by: Lisa Seelye <lisa@users.noreply.github.com>

lisa committed:
Fix raci-chart link (kind of)

It's not really fixed in so much as it's not pointing to somewhere that's just blatantly wrong now.
